### PR TITLE
[ALLUXIO-2182]  Display per-UFS IO metrics

### PR DESCRIPTION
**Pull request by yupeng9**
https://alluxio.atlassian.net/browse/ALLUXIO-2182

Displays the IO metrics per under file system.

### CODE REVIEW

**Comment by calvinjia** [comment]
@yupeng9 Based on @aaudiber's suggestion, we will change metrics system from a singleton, can you hold off on this until #7391 is resolved?

**Comment by yupeng9** [comment]
@calvinjia SGTM

File: core/common/src/main/java/alluxio/metrics/MetricsSystem.java [hunk 1]
```diff
@@ -56,7 +56,8 @@
   public enum InstanceType {
     MASTER("master"),
     WORKER("worker"),
-    CLIENT("client");
+    CLIENT("client"),
+    CLUSTER("cluster");
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/metrics/MetricsSystem.java]: Does this mean we can get rid of the String constant?

File: core/common/src/main/java/alluxio/metrics/aggregator/SingleTagValueAggregator.java [hunk 2]
```diff
@@ -68,7 +68,7 @@ public SingleTagValueAggregator(MetricsSystem.InstanceType instanceType, String
       Map<String, String> tags = metric.getTags();
       if (tags.containsKey(mTagName)) {
         String ufsName = MetricsSystem.getClusterMetricName(
-            mMetricName + "." + tags.get(mTagName));
+            mMetricName + "." + mTagName + Metric.TAG_SEPARATOR + tags.get(mTagName));
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/metrics/aggregator/SingleTagValueAggregator.java]: Should we be using `Metric.getMetricNameWithTags`?

File: core/server/master/src/main/java/alluxio/web/WebInterfaceMasterMetricsServlet.java [hunk 3]
```diff
@@ -129,28 +132,40 @@ public boolean matches(String name, Metric metric) {
...
+        .get(MetricsSystem.getClusterMetricName(WorkerMetrics.BYTES_READ_ALLUXIO)).getValue();
     Long bytesReadUfs = (Long) mr.getGauges()
-        .get(MetricsSystem.getClusterMetricName(WorkerMetrics.BYTES_READ_UFS_ALL))
-        .getValue();
+        .get(MetricsSystem.getClusterMetricName(WorkerMetrics.BYTES_READ_UFS_ALL)).getValue();
     request.setAttribute("totalBytesReadLocal", FormatUtils.getSizeFromBytes(bytesReadLocal));
     request.setAttribute("totalBytesReadRemote", FormatUtils.getSizeFromBytes(bytesReadRemote));
     request.setAttribute("totalBytesReadUfs", FormatUtils.getSizeFromBytes(bytesReadUfs));
 
+    // cluster cache hit and miss
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/web/WebInterfaceMasterMetricsServlet.java]: Can we add cache hit ratio per tier?

File: core/common/src/main/java/alluxio/metrics/MetricsSystem.java (hunk 1 above)
Comment (yupeng9) [reply core/common/src/main/java/alluxio/metrics/MetricsSystem.java]: We used the constant for string comparison without transformation.

File: core/common/src/main/java/alluxio/metrics/aggregator/SingleTagValueAggregator.java (hunk 2 above)
Comment (yupeng9) [reply core/common/src/main/java/alluxio/metrics/aggregator/SingleTagValueAggregator.java]: done

File: core/server/master/src/main/java/alluxio/web/WebInterfaceMasterMetricsServlet.java (hunk 3 above)
Comment (yupeng9) [reply core/server/master/src/main/java/alluxio/web/WebInterfaceMasterMetricsServlet.java]: Will add a followup commit for this.

**Comment by yupeng9** [comment]
@calvinjia PTAL

File: core/common/src/main/java/alluxio/metrics/MetricsSystem.java (hunk 1 above)
Comment (calvinjia) [reply core/common/src/main/java/alluxio/metrics/MetricsSystem.java]: Could we not use `InstanceType.CLUSTER.toString()`?

**Comment by calvinjia** [comment]
@yupeng9 Had a follow up question, also does `valueOf` not work in place of `fromString` for the enum?

**Comment by yupeng9** [comment]
@calvinjia It could with the `toLowerCase` transformation. I don't have a preference on either, so I'll change it to what you requested.

File: core/common/src/main/java/alluxio/metrics/MetricsSystem.java [hunk 4]
```diff
@@ -55,25 +55,17 @@
...
      */
-    InstanceType(String value) {
-      mValue = value;
-    }
-
-    @Override
-    public String toString() {
-      return mValue;
+    public String getString() {
+      return this.toString().toLowerCase();
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/metrics/MetricsSystem.java]: How important is it to do this in lower case?

File: core/common/src/main/java/alluxio/metrics/WorkerMetrics.java [hunk 5]
```diff
@@ -15,6 +15,34 @@
...
+    GetStatus,
+    IsDirectory,
+    IsFile,
+    ListStatus,
+    Mkdirs,
+    Open,
+    RenameDirectory,
+    RenameFile,
+    SetOwner,
+    SetMode;
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/metrics/WorkerMetrics.java]: Should these be in caps? Also this is a fairly hard to maintain list of operations, can we determine whether or not it is a ufs op from the metric name itself instead of having to match this list?

File: core/server/master/src/main/java/alluxio/master/metrics/DefaultMetricsMaster.java [hunk 6]
```diff
@@ -146,10 +146,15 @@ private void registerAggregators() {
...
     // multi-value aggregators
-    addAggregator(new SingleTagValueAggregator(MetricsSystem.InstanceType.WORKER,
-        WorkerMetrics.BYTES_READ_UFS, WorkerMetrics.TAG_UFS));
-    addAggregator(new SingleTagValueAggregator(MetricsSystem.InstanceType.WORKER,
-        WorkerMetrics.BYTES_WRITTEN_UFS, WorkerMetrics.TAG_UFS));
+    addAggregator(new SingleTagValueAggregator(WorkerMetrics.BYTES_READ_UFS,
+        MetricsSystem.InstanceType.WORKER, WorkerMetrics.BYTES_READ_UFS, WorkerMetrics.TAG_UFS));
+    addAggregator(new SingleTagValueAggregator(WorkerMetrics.BYTES_WRITTEN_UFS,
+        MetricsSystem.InstanceType.WORKER, WorkerMetrics.BYTES_WRITTEN_UFS, WorkerMetrics.TAG_UFS));
+    for (WorkerMetrics.UfsOps ufsOp : WorkerMetrics.UfsOps.values()) {
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/metrics/DefaultMetricsMaster.java]: Could we add these on the fly as we get them, then we would not have to keep track of a master list.

File: core/server/master/src/main/java/alluxio/master/metrics/DefaultMetricsMaster.java [hunk 7]
```diff
@@ -146,10 +146,15 @@ private void registerAggregators() {
...
+      addAggregator(new SingleTagValueAggregator(WorkerMetrics.UFS_OP_PREFIX + ufsOp,
+          MetricsSystem.InstanceType.MASTER, ufsOp.toString(),
+          WorkerMetrics.TAG_UFS));
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/metrics/DefaultMetricsMaster.java]: This aggregator is different from the throughput ones which are separated per mount point. For example, if there are two s3 mount, we still only keep track of one s3 metric.

File: core/common/src/main/java/alluxio/metrics/MetricsSystem.java (hunk 4 above)
Comment (yupeng9) [reply core/common/src/main/java/alluxio/metrics/MetricsSystem.java]: This value will be part of the metric name, and for now the convention is to use lower case?

File: core/common/src/main/java/alluxio/metrics/WorkerMetrics.java (hunk 5 above)
Comment (yupeng9) [reply core/common/src/main/java/alluxio/metrics/WorkerMetrics.java]: I cannot think of a good way of getting the metric names. It is used when the default metric master starts, and the reflection may not work well because there are other methods on the under filesystem interface not accounted as ufs ops.

File: core/server/master/src/main/java/alluxio/master/metrics/DefaultMetricsMaster.java (hunk 6 above)
Comment (yupeng9) [reply core/server/master/src/main/java/alluxio/master/metrics/DefaultMetricsMaster.java]: hmm how to do it?

File: core/server/master/src/main/java/alluxio/master/metrics/DefaultMetricsMaster.java (hunk 7 above)
Comment (yupeng9) [reply core/server/master/src/main/java/alluxio/master/metrics/DefaultMetricsMaster.java]: done

File: core/server/master/src/main/java/alluxio/master/metrics/DefaultMetricsMaster.java (hunk 6 above)
Comment (calvinjia) [reply core/server/master/src/main/java/alluxio/master/metrics/DefaultMetricsMaster.java]: I was thinking of doing something like prefixing all the ops with `UfsOp` and then if a metric comes in with the prefix we would create an aggregator for it if it has not been created. But this approach is also not clean, and I'm okay with the enum if you think that is preferable.

**Comment by calvinjia** [comment]
@yupeng9 Looks good overall, could you update what the new metrics page looks like with the new aggregation?

File: core/server/master/src/main/java/alluxio/master/metrics/DefaultMetricsMaster.java (hunk 6 above)
Comment (yupeng9) [reply core/server/master/src/main/java/alluxio/master/metrics/DefaultMetricsMaster.java]: Right. I think a central place for registration offers better modularity and code cleanness. So I feel it's worth maintaining an extra enum data structure

**Comment by yupeng9** [comment]
Sure

**Comment by calvinjia** [comment]
Jenkins, test this please.

**Comment by calvinjia** [comment]
@yupeng9 Do you know if the test failures are related to your changes?

**Comment by yupeng9** [comment]
@calvinjia that's the same flakiness. I turned the client-side metrics collection on in this PR.

**Comment by yupeng9** [comment]
jenkins, test this please

**Comment by yupeng9** [comment]
jenkins, test this please

**Comment by yupeng9** [comment]
jenkins, test this please

**Comment by yupeng9** [comment]
jenkins, test this please

**Comment by yupeng9** [comment]
jenkins, test this please

**Comment by yupeng9** [comment]
jenkins, test this please

**Comment by yupeng9** [comment]
jenkins, test this please

**Comment by yupeng9** [comment]
jenkins, test this please

**Comment by yupeng9** [comment]
@calvinjia @aaudiber rewrote the history and retargeted to branch-1.8

**Comment by yupeng9** [comment]
jenkins, test this please

**Comment by yupeng9** [comment]
jenkins, test this please

**Comment by yupeng9** [comment]
jenkins, test this please

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 8]
```diff
@@ -177,17 +178,18 @@ private FileSystemContext(Subject subject) {
    * Initializes the context. Only called in the factory methods and reset.
    *
    * @param masterInquireClient the client to use for determining the master
+   * @param enableUserMetricsCollection if enables the user metrics collection
    */
-  private synchronized void init(MasterInquireClient masterInquireClient) {
+  private synchronized void init(MasterInquireClient masterInquireClient,
+      boolean enableUserMetricsCollection) {
```
Comment (aaudiber) [inline core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: instead of passing `enableUserMetricsCollection` as a boolean, could callers just set the Configuration property?

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 9]
```diff
@@ -420,6 +420,7 @@ public void setPermission(Path path, FsPermission permission) throws IOException
    *      org.apache.hadoop.conf.Configuration)
    */
   //@Override This doesn't exist in Hadoop 1.x, so cannot put {@literal @Override}.
+  @Override
```
Comment (aaudiber) [inline core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: see comment on above line

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java (hunk 8 above)
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: hmm, I thought about that. The difference is that having a parameter here notifies the caller that this configuration should be considered.

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 10]
```diff
@@ -178,18 +178,19 @@ private FileSystemContext(Subject subject) {
...
     mMasterInquireClient = masterInquireClient;
     mFileSystemMasterClientPool =
         new FileSystemMasterClientPool(mParentSubject, mMasterInquireClient);
     mBlockMasterClientPool = new BlockMasterClientPool(mParentSubject, mMasterInquireClient);
     mClosed.set(false);
 
     // Only send metrics if enabled and the port is set (can be zero when tests are setting up).
-    if (enableUserMetricsCollection && Configuration.getInt(PropertyKey.MASTER_RPC_PORT) != 0) {
+    if (configuration.getBoolean(PropertyKey.USER_METRICS_COLLECTION_ENABLED)
+        && Configuration.getInt(PropertyKey.MASTER_RPC_PORT) != 0) {
```
Comment (aaudiber) [inline core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: can we remove the `Configuration.getInt(PropertyKey.MASTER_RPC_PORT) != 0` now?

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java [hunk 11]
```diff
@@ -178,18 +178,19 @@ private FileSystemContext(Subject subject) {
    * Initializes the context. Only called in the factory methods and reset.
    *
    * @param masterInquireClient the client to use for determining the master
-   * @param enableUserMetricsCollection if enables the user metrics collection
+   * @param configuration the instance configuration
    */
   private synchronized void init(MasterInquireClient masterInquireClient,
-      boolean enableUserMetricsCollection) {
+      InstancedConfiguration configuration) {
```
Comment (aaudiber) [inline core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: use this configuration instead of `Configuration.getMs(PropertyKey.USER_METRICS_HEARTBEAT_INTERVAL_MS)` below?

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java (hunk 10 above)
Comment (yupeng9) [reply core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java]: yes, good call

**Comment by yupeng9** [comment]
@calvinjia @aaudiber any more comments?

File: core/common/src/main/java/alluxio/metrics/Metric.java [hunk 12]
```diff
@@ -245,8 +245,7 @@ public static Metric from(String fullName, double value) {
     for (int i = tagStartIdx; i < pieces.length; i++) {
       String tagStr = pieces[i];
       if (!tagStr.contains(TAG_SEPARATOR)) {
-        LOG.warn("Unknown tag {}, the tag format should be tagname{}tagvalue", pieces[i],
-            TAG_SEPARATOR);
+        // Unknown tag
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/metrics/Metric.java]: Should we log a warning still?

File: core/common/src/main/java/alluxio/underfs/UnderFileSystem.java [hunk 13]
```diff
@@ -99,7 +99,7 @@ public static UnderFileSystem create(String path, UnderFileSystemConfiguration u
           // when creation is done.
           Thread.currentThread().setContextClassLoader(factory.getClass().getClassLoader());
           // Use the factory to create the actual client for the Under File System
-          return new UnderFileSystemWithLogging(factory.create(path, ufsConf));
+          return new UnderFileSystemWithLogging(path, factory.create(path, ufsConf));
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystem.java]: Is it guaranteed that `path` is the mount path?

File: core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java [hunk 14]
```diff
@@ -779,7 +779,7 @@ public Command workerHeartbeat(long workerId, Map<String, Long> usedBytesOnTiers
       return new Command(CommandType.Register, new ArrayList<Long>());
     }
 
-    synchronized (worker) {
+    synchronized (worker.getWorkerAddress()) {
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java]: Why is this changed?

File: core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java [hunk 15]
```diff
@@ -139,9 +124,33 @@ public synchronized void putClientMetrics(String hostname, String clientId,
...
+    if (instanceType == InstanceType.WORKER) {
+      synchronized (mWorkerMetrics) {
+        return mWorkerMetrics.getByField(NAME_INDEX, name);
+      }
+    } else if (instanceType == InstanceType.CLIENT) {
+      synchronized (mWorkerMetrics) {
+        return mClientMetrics.getByField(NAME_INDEX, name);
+      }
+    } else {
+      throw new IllegalArgumentException("Unsupported instance type " + instanceType);
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java]: Should `PROXY` and `CLUSTER` be supported?

File: core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java (hunk 14 above)
Comment (aaudiber) [reply core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java]: This is really hacky now, re-purposing the mWorkerAddress field as an alternate way to lock the WorkerInfo. How are developers supposed to know when to synchronize on worker vs when to synchronize on `worker.getWorkerAddress()`? And why the worker address? I don't think this is a reasonable way to implement the locking here

File: core/common/src/main/java/alluxio/metrics/Metric.java (hunk 12 above)
Comment (yupeng9) [reply core/common/src/main/java/alluxio/metrics/Metric.java]: There will be excessive warning loggings, as some metrics (e.g. JVM's) do not conform this convention.

**Comment by yupeng9** [comment]
@calvinjia, yes, `USER_METRICS_COLLECTION_ENABLED` is disabled only for 1 test.

In addition, client-side metric collection is tested in test `MetricsMasterTest`.

File: core/common/src/main/java/alluxio/underfs/UnderFileSystem.java (hunk 13 above)
Comment (yupeng9) [reply core/common/src/main/java/alluxio/underfs/UnderFileSystem.java]: The path is the UFS path, but not Alluxio's mounting path.

Comment (calvinjia) [reply core/common/src/main/java/alluxio/underfs/UnderFileSystem.java]: Right, is this going to be whichever UFS path that gets accessed first, or will it specifically be the mounted UFS path (ie. `hdfs:///data` vs `hdfs:///data/2018/06/19/myfile.parquet`)?

**Comment by calvinjia** [comment]
@yupeng9 Aren't we disabling `USER_METRICS_COLLECTION_ENABLED` in `AbstractLocalAlluxioCluster.java`?

File: core/common/src/main/java/alluxio/metrics/Metric.java (hunk 12 above)
Comment (calvinjia) [reply core/common/src/main/java/alluxio/metrics/Metric.java]: We currently use this method to serialize metrics to send to the master. Can we only send the metrics which make sense to send (ie. the Alluxio generated ones which we will aggregate at the cluster level)?

File: core/common/src/main/java/alluxio/underfs/UnderFileSystem.java (hunk 13 above)
Comment (yupeng9) [reply core/common/src/main/java/alluxio/underfs/UnderFileSystem.java]: the former like `hdfs:///data`.

This method is called when a UFS registers, i.e. a new UFS mounts. So the path will be the url that corresponds to the mounting point.

File: core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java (hunk 15 above)
Comment (yupeng9) [reply core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java]: For now we do not collect and aggregate proxy metrics. And `cluster` metrics are aggregated results, so they are not handled here.

**Comment by yupeng9** [comment]
@calvinjia That's a good call. `USER_METRICS_COLLECTION_ENABLED` should not be set in  `AbstractLocalAlluxioCluster.java.

**Comment by yupeng9** [comment]
@calvinjia PTAL.

File: core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java (hunk 9 above)
Comment (aaudiber) [reply core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: this slipped through and broke the hadoop-1 build when I tried to generate RC3 artifacts
